File: java/sorting/walkthrough/QuickSort.java
/*
Quick Sort Algorithm

Quick Sort is a divide and conquer sorting algorithm.

How it works:
- It selects a pivot element from the list.
- It partitions the remaining elements into three groups.
- Elements smaller than the pivot go to the left group.
- Elements equal to the pivot stay in the middle group.
- Elements greater than the pivot go to the right group.
- The left and right groups are sorted recursively.
- The final result is the combination of left, middle, and right groups.

This algorithm is usually very fast in practice.

Time Complexity:
- Worst Case: O(n^2)
- Average Case: O(n log n)
- Best Case: O(n log n)

Space Complexity:
- O(n) in this implementation due to extra lists.

Stability:
- Not stable in general
*/

import java.util.*;

public class QuickSort {

    public static int[] quickSortVerbose(int[] arr) {
        return quickSortVerboseRecursive(arr, 0);
    }

    private static int[] quickSortVerboseRecursive(int[] arr, int depth) {

        String indent = new String(new char[depth]).replace("\0", "  ");
        System.out.println(indent + "quick_sort(" + Arrays.toString(arr) + ")");

        if (arr.length <= 1) {
            return Arrays.copyOf(arr, arr.length);
        }

        int pivot = arr[arr.length / 2];
        System.out.println(indent + "Pivot: " + pivot);

        List<Integer> left = new ArrayList<>();
        List<Integer> middle = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        for (int x : arr) {

            System.out.println(indent + "Compare " + x + " with pivot " + pivot);

            if (x < pivot) {
                left.add(x);
                System.out.println(indent + "  -> go left: " + left);
            } else if (x > pivot) {
                right.add(x);
                System.out.println(indent + "  -> go right: " + right);
            } else {
                middle.add(x);
                System.out.println(indent + "  -> go middle: " + middle);
            }
        }

        int[] sortedLeft = quickSortVerboseRecursive(
                left.stream().mapToInt(i -> i).toArray(),
                depth + 1
        );

        int[] sortedRight = quickSortVerboseRecursive(
                right.stream().mapToInt(i -> i).toArray(),
                depth + 1
        );

        int[] result = concatenate(
                sortedLeft,
                middle.stream().mapToInt(i -> i).toArray(),
                sortedRight
        );

        System.out.println(indent + "Combined result: "
                + Arrays.toString(result));

        return result;
    }

    private static int[] concatenate(int[] left, int[] middle, int[] right) {

        int[] result = new int[left.length + middle.length + right.length];

        int index = 0;

        for (int value : left) {
            result[index++] = value;
        }

        for (int value : middle) {
            result[index++] = value;
        }

        for (int value : right) {
            result[index++] = value;
        }

        return result;
    }
}

File: python/sorting/walkthrough/quick_sort.py
"""
Quick Sort Algorithm

Quick Sort is a divide and conquer sorting algorithm.

How it works:
- It selects a pivot element from the list.
- It partitions the remaining elements into three groups.
- Elements smaller than the pivot go to the left group.
- Elements equal to the pivot stay in the middle group.
- Elements greater than the pivot go to the right group.
- The left and right groups are sorted recursively.
- The final result is the combination of left, middle, and right groups.

This algorithm is usually very fast in practice.

Time Complexity:
- Worst Case: O(n^2)
- Average Case: O(n log n)
- Best Case: O(n log n)

Space Complexity:
- O(n) in this implementation due to extra lists.

Stability:
- Not stable in general
"""


def quick_sort_verbose(arr, depth=0):
    indent = "  " * depth
    print(f"{indent}quick_sort({arr})")

    if len(arr) <= 1:
        return arr[:]

    pivot = arr[len(arr) // 2]
    print(f"{indent}Pivot: {pivot}")

    left = []
    middle = []
    right = []

    for value in arr:
        print(f"{indent}Compare {value} with pivot {pivot}")
        if value < pivot:
            left.append(value)
            print(f"{indent}  -> go left: {left}")
        elif value > pivot:
            right.append(value)
            print(f"{indent}  -> go right: {right}")
        else:
            middle.append(value)
            print(f"{indent}  -> go middle: {middle}")

    sorted_left = quick_sort_verbose(left, depth + 1)
    sorted_right = quick_sort_verbose(right, depth + 1)

    result = sorted_left + middle + sorted_right
    print(f"{indent}Combined result: {result}")
    return result
